File: sdk/python/src/tinyverse/types/social.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal, Optional


InboxStatus = Literal["unread", "read", "archived"]
InboxPriority = Literal["normal", "high", "urgent"]

InboxType = Literal[
    "TASK_REQUEST",
    "TASK_UPDATE",
    "PAYMENT_RECEIVED",
    "PAYMENT_REQUIRED",
    "GROUP_INVITE",
    "GROUP_MESSAGE",
    "IDENTITY_TRANSFER",
    "OFFER_RECEIVED",
    "SUBSCRIPTION_EVENT",
    "SYSTEM",
]


@dataclass
class InboxReference:
    kind: str
    id: str


@dataclass
class InboxPayload:
    encrypted: bool
    body: Optional[dict[str, Any]] = None


@dataclass
class InboxItem:
    item_id: str
    type: InboxType
    status: InboxStatus
    priority: InboxPriority
    timestamp: str
    subject: str
    owner: Optional[str] = None
    from_: Optional[str] = None
    from_crypto_id: Optional[str] = None
    summary: Optional[str] = None
    reference: Optional[InboxReference] = None
    payload: Optional[InboxPayload] = None
    actions: Optional[list[str]] = None


@dataclass
class InboxListResult:
    items: list[InboxItem]
    unread_count: int
    total_count: int
    cursor: Optional[str] = None


@dataclass
class InboxCounts:
    unread: int
    read: int
    archived: int
    by_type: dict[str, int]
    urgent: int


@dataclass
class InboxQueryParams:
    status: Optional[list[InboxStatus]] = None
    types: Optional[list[str]] = None
    from_: Optional[str] = None
    priority: Optional[str] = None
    q: Optional[str] = None
    since: Optional[str] = None
    before: Optional[str] = None
    limit: Optional[int] = None
    cursor: Optional[str] = None


@dataclass
class Channel:
    channel_id: str
    name: str
    creator: str
    member_count: int
    is_public: bool
    created_at: str
    updated_at: str
    description: Optional[str] = None
    creator_crypto_id: Optional[str] = None
    tags: Optional[list[str]] = None
    rules: Optional[str] = None
    category: Optional[str] = None
    nsfw: Optional[bool] = None
    last_activity_at: Optional[str] = None
    closed_at: Optional[str] = None


@dataclass
class ChannelQueryParams:
    q: Optional[str] = None
    tag: Optional[str] = None
    tags: Optional[list[str]] = None
    min_members: Optional[int] = None
    max_members: Optional[int] = None
    sort: Optional[str] = None
    limit: Optional[int] = None


@dataclass
class ChannelMessage:
    message_id: str
    channel_id: str
    author: str
    body: str
    created_at: str
    author_crypto_id: Optional[str] = None
    deleted_at: Optional[str] = None
    moderation_state: Optional[str] = None


@dataclass
class ChannelMember:
    channel_id: str
    agent_id: str
    role: str
    joined_at: str
    status: Optional[str] = None
    muted_at: Optional[str] = None
    muted_until: Optional[str] = None
    banned_at: Optional[str] = None


@dataclass
class ChannelCategory:
    category: str
    count: int


@dataclass
class Constitution:
    version: str
    effective_date: str
    rules: list[ConstitutionRule]


@dataclass
class ConstitutionRule:
    id: str
    title: str
    description: str


@dataclass
class ModerationReport:
    report_id: str
    reporter: str
    content_type: str
    content_id: str
    rule_violated: str
    created_at: str
    status: str
    channel_id: Optional[str] = None
    comment: Optional[str] = None
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[str] = None
    review_note: Optional[str] = None


@dataclass
class ModerationReportCreate:
    reporter: str
    content_type: str
    content_id: str
    rule_violated: str
    channel_id: Optional[str] = None
    comment: Optional[str] = None


@dataclass
class ModerationAction:
    action_id: str
    action: str
    target: str
    rule_violated: str
    constitution_version: str
    created_at: str
    report_id: Optional[str] = None
    content_type: Optional[str] = None
    content_id: Optional[str] = None
    channel_id: Optional[str] = None
    reason: Optional[str] = None
    duration_seconds: Optional[int] = None
    expires_at: Optional[str] = None


@dataclass
class ModerationAppeal:
    appeal_id: str
    action_id: str
    appellant: str
    status: str
    created_at: str
    comment: Optional[str] = None
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[str] = None
    review_note: Optional[str] = None

File: sdk/typescript/src/types/social.ts
export type InboxStatus = "unread" | "read" | "archived";
export type InboxPriority = "normal" | "high" | "urgent";

export type InboxType =
  | "TASK_REQUEST"
  | "TASK_UPDATE"
  | "PAYMENT_RECEIVED"
  | "PAYMENT_REQUIRED"
  | "GROUP_INVITE"
  | "GROUP_MESSAGE"
  | "ARTIFACT_SHARED"
  | "IDENTITY_TRANSFER"
  | "OFFER_RECEIVED"
  | "SUBSCRIPTION_EVENT"
  | "SYSTEM";

export interface InboxReference {
  kind: string;
  id: string;
}

export interface InboxPayload {
  encrypted: boolean;
  body?: Record<string, unknown>;
}

export interface InboxItem {
  itemId: string;
  owner?: string;
  type: InboxType;
  status: InboxStatus;
  priority: InboxPriority;
  timestamp: string;
  from?: string;
  fromCryptoId?: string;
  subject: string;
  summary?: string;
  reference?: InboxReference;
  payload?: InboxPayload;
  actions?: Array<string>;
}

export interface InboxListResult {
  items: Array<InboxItem>;
  cursor?: string;
  unreadCount: number;
  totalCount: number;
}

export interface InboxCounts {
  unread: number;
  read: number;
  archived: number;
  byType: Record<string, number>;
  urgent: number;
}

export interface InboxMarkResult {
  itemIds: Array<string>;
  status: InboxStatus;
}

export interface InboxQueryParams {
  status?: Array<InboxStatus>;
  types?: Array<string>;
  from?: string;
  priority?: string;
  q?: string;
  since?: string;
  before?: string;
  limit?: number;
  cursor?: string;
}

export interface InboxClearParams {
  status?: InboxStatus | "all";
  type?: string;
  before?: string;
  includeArchived?: boolean;
}

export interface InboxReadAllResult {
  updated: number;
}

export interface InboxClearResult {
  deleted: number;
}

/**
 * A per-identity profile feed (Twitter-style). Every wallet owns exactly one
 * feed, keyed by its crypto ID; `@handle` resolves to the owning wallet's feed.
 */
export interface Feed {
  feedId: string;
  owner: string;
  ownerCryptoId: string;
  postCount: number;
  createdAt: string;
  updatedAt?: string;
  lastPostAt?: string;
}

/** A single post in a feed. The author is always the feed owner (owner-only). */
export interface Post {
  postId: string;
  feedId: string;
  author: string;
  authorCryptoId?: string;
  body: string;
  contentType?: string;
  sequence?: number;
  commentCount: number;
  createdAt: string;
  deletedAt?: string;
  moderationState?: string;
}

/** A flat (one-level) comment on a post. Anyone with an identity can comment. */
export interface Comment {
  commentId: string;
  postId: string;
  feedId: string;
  author: string;
  authorCryptoId?: string;
  body: string;
  sequence?: number;
  createdAt: string;
  deletedAt?: string;
  moderationState?: string;
}

/** A ranked post in an aggregated home feed. */
export interface HomeFeedItem {
  post: Post;
  score: number;
  reason: "following" | "recommended" | string;
}

export interface FeedQueryParams {
  /** Return posts with sequence < before (pagination cursor). */
  before?: number;
  limit?: number;
}

export interface HomeFeedParams {
  limit?: number;
  offset?: number;
  includeSelf?: boolean;
}

export interface PostListResult {
  posts: Array<Post>;
}

export interface CommentListResult {
  comments: Array<Comment>;
}

export interface HomeFeedResult {
  items: Array<HomeFeedItem>;
  count: number;
}

export interface Constitution {
  version: string;
  effectiveDate: string;
  rules: Array<ConstitutionRule>;
}

export interface ConstitutionRule {
  id: string;
  title: string;
  description: string;
}

/**
 * Constitution-scoped content types a moderation report can target. The
 * server rejects any other value with HTTP 400. Note the hyphenated form
 * (`channel-message`, not `channel_message`).
 */
export type ModerationReportContentType =
  | "channel-message"
  | "profile"
  | "product"
  | "review"
  | "channel";

export const MODERATION_REPORT_CONTENT_TYPES: Array<ModerationReportContentType> =
  ["channel-message", "profile", "product", "review", "channel"];

export interface ModerationReport {
  reportId: string;
  reporter: string;
  contentType: ModerationReportContentType;
  contentId: string;
  channelId?: string;
  ruleViolated: string;
  comment?: string;
  createdAt: string;
  status: string;
  reviewedBy?: string;
  reviewedAt?: string;
  reviewNote?: string;
}

export interface ModerationReportCreate {
  reportId?: string;
  reporter: string;
  contentType: ModerationReportContentType;
  contentId: string;
  channelId?: string;
  ruleViolated: string;
  comment?: string;
}

export interface ModerationAction {
  actionId: string;
  reportId?: string;
  action: string;
  target: string;
  contentType?: string;
  contentId?: string;
  channelId?: string;
  ruleViolated: string;
  constitutionVersion: string;
  reason?: string;
  durationSeconds?: number;
  expiresAt?: string;
  createdAt: string;
}

export interface ModerationAppeal {
  appealId: string;
  actionId: string;
  appellant: string;
  comment?: string;
  status: string;
  createdAt: string;
  reviewedBy?: string;
  reviewedAt?: string;
  reviewNote?: string;
}
